Images always filling space after 2.2.3 - IMPORTANT!!! CachedImage.FixedOnMeasureBehavior property

**//EDIT by Daniel Luberda:**

It's fixed now, but you must set this:

`CachedImage.FixedOnMeasureBehavior = true;`

**I had to do this to avoid breaking behavior changes as the regression exists now for quite a few versions. We'll reenable it with next major version to avoid confusion.** 

//ORIGINAL: 

Hi,

Before I go to the effort of providing a lot of detail on this, I wanted to put feelers out first. I've been hanging back on 2.2.3 for some time now because when I upgrade, my images in XF appear way too big. I assumed this was a bug in FFIL when I first upgraded (I believe it was 2.2.4), but since it's been that way for a _long_ time now, I figure maybe this is intentional?

Was there some kind of default behavior change at this point? I _think_ this only affects Android btw, but would have to confirm that as part of digging deeper if/when required.

This is the build error I get, btw:
```
9>C:\Users\Kent\Repository\FFImageLoading\source\FFImageLoading.BaitAndSwitch\project.json(1,1): error MSB4025: The project file could not be loaded. Data at the root level is invalid. Line 1, position 1.
```
Can't see anything obviously wrong with the file in question (and I haven't changed anything since cloning the repo).

Hmm, it's still using project.json, which is leading to this when I try and use FFIL source directly from my Xamarin project:
```
C:\Users\Kent\Repository\FFImageLoading\source\FFImageLoading.Common\project.json(1,1): error MSB4025: The project file could not be loaded. Data at the root level is invalid. Line 1, position 1.
```
Ignoring that, the good news is that I can now build FFIL in isolation from my Xamarin project. So I figured i could just reference the compiled binaries from my solution instead of using FFIL source. However, when referencing those binaries from my Xamarin project, my XAML compilation started failing (with null ref exceptions). It's not clear why, so I just disabled XAMLC to continue on.

Basically, it's really quite difficult working with FFIL as source, and I suspect the friction prevents quite a few community contributions. You should take a look at what the ReactiveUI folks (mainly @ollevett) have done to multitarget many platforms from one project. You'd wind up with 3 projects instead of `3 * $PLATFORM_COUNT`. Anyway, just a thought - you may not have the time to invest right now - completely understand if that's the case.

That said, I finally got building (against the binaries, so not a great debugging story). However, the problem is still occurring, and debugging is not really practical without source :(

It's really simple to reproduce. You just give an image more space than it needs and set it to center. A standard `Image` does the right thing, but `CachedImage` stretches the image to fill the space. Again, this regressed some time after 2.2.3.

[Here's a repro](https://github.com/luberda-molinet/FFImageLoading/files/1273897/FFILRepro.zip).

--- a/source/FFImageLoading.Forms/CachedImage.cs
+++ b/source/FFImageLoading.Forms/CachedImage.cs
@@ -28,7 +28,7 @@ namespace FFImageLoading.Forms
 		/// <summary>
 		/// Gets or sets the aspect.
 		/// </summary>
-		/// <value>The aspect.</value> 
+		/// <value>The aspect.</value>
 		public Aspect Aspect
 		{
 			get
@@ -86,17 +86,17 @@ namespace FFImageLoading.Forms
 
         /// <summary>
         /// The source property.
-        /// </summary> 
+        /// </summary>
         public static readonly BindableProperty SourceProperty = BindableProperty.Create(nameof (Source), typeof(ImageSource), typeof(CachedImage), default(ImageSource), BindingMode.OneWay, propertyChanged: OnSourcePropertyChanged);
 
         static void OnSourcePropertyChanged(BindableObject bindable, object oldValue, object newValue)
         {
-            if (newValue != null) 
+            if (newValue != null)
             {
                 BindableObject.SetInheritedBindingContext(newValue as BindableObject, bindable.BindingContext);
             }
         }
-			
+
 		/// <summary>
 		/// Gets or sets the source.
 		/// </summary>
@@ -126,11 +126,11 @@ namespace FFImageLoading.Forms
 		{
 			get
 			{
-				return (int)GetValue(RetryCountProperty); 
+				return (int)GetValue(RetryCountProperty);
 			}
 			set
 			{
-				SetValue(RetryCountProperty, value); 
+				SetValue(RetryCountProperty, value);
 			}
 		}
 
@@ -146,11 +146,11 @@ namespace FFImageLoading.Forms
 		{
 			get
 			{
-				return (int)GetValue(RetryDelayProperty); 
+				return (int)GetValue(RetryDelayProperty);
 			}
 			set
 			{
-				SetValue(RetryDelayProperty, value); 
+				SetValue(RetryDelayProperty, value);
 			}
 		}
 
@@ -180,18 +180,18 @@ namespace FFImageLoading.Forms
         public static readonly BindableProperty DownsampleWidthProperty = BindableProperty.Create(nameof(DownsampleWidth), typeof(double), typeof(CachedImage), 0d);
 
 		/// <summary>
-		/// Reduce memory usage by downsampling the image. Aspect ratio will be kept even if width/height values are incorrect. 
+		/// Reduce memory usage by downsampling the image. Aspect ratio will be kept even if width/height values are incorrect.
 		/// Optional DownsampleWidth parameter, if value is higher than zero it will try to downsample to this width while keeping aspect ratio.
 		/// </summary>
 		public double DownsampleWidth
 		{
 			get
 			{
-				return (double)GetValue(DownsampleWidthProperty); 
+				return (double)GetValue(DownsampleWidthProperty);
 			}
 			set
 			{
-				SetValue(DownsampleWidthProperty, value); 
+				SetValue(DownsampleWidthProperty, value);
 			}
 		}
 
@@ -201,18 +201,18 @@ namespace FFImageLoading.Forms
         public static readonly BindableProperty DownsampleHeightProperty = BindableProperty.Create(nameof(DownsampleHeight), typeof(double), typeof(CachedImage), 0d);
 
 		/// <summary>
-		/// Reduce memory usage by downsampling the image. Aspect ratio will be kept even if width/height values are incorrect. 
+		/// Reduce memory usage by downsampling the image. Aspect ratio will be kept even if width/height values are incorrect.
 		/// Optional DownsampleHeight parameter, if value is higher than zero it will try to downsample to this height while keeping aspect ratio.
 		/// </summary>
 		public double DownsampleHeight
 		{
 			get
 			{
-				return (double)GetValue(DownsampleHeightProperty); 
+				return (double)GetValue(DownsampleHeightProperty);
 			}
 			set
 			{
-				SetValue(DownsampleHeightProperty, value); 
+				SetValue(DownsampleHeightProperty, value);
 			}
 		}
 
@@ -224,7 +224,7 @@ namespace FFImageLoading.Forms
 		/// <summary>
 		/// Reduce memory usage by downsampling the image. Aspect ratio will be kept even if width/height values are incorrect.
 		/// DownsampleWidth and DownsampleHeight properties will be automatically set to view size
-		/// If the view height or width will not return > 0 - it'll fall back 
+		/// If the view height or width will not return > 0 - it'll fall back
 		/// to using DownsampleWidth / DownsampleHeight properties values
 		/// IMPORTANT: That property is tricky when using some auto-layouts as view doesn't have its size defined,
 		/// so it's always safe to have DownsampleWidth / DownsampleHeight set as a fallback
@@ -234,11 +234,11 @@ namespace FFImageLoading.Forms
 		{
 			get
 			{
-				return (bool)GetValue(DownsampleToViewSizeProperty); 
+				return (bool)GetValue(DownsampleToViewSizeProperty);
 			}
 			set
 			{
-				SetValue(DownsampleToViewSizeProperty, value); 
+				SetValue(DownsampleToViewSizeProperty, value);
 			}
 		}
 
@@ -248,7 +248,7 @@ namespace FFImageLoading.Forms
         public static readonly BindableProperty DownsampleUseDipUnitsProperty = BindableProperty.Create(nameof(DownsampleUseDipUnits), typeof(bool), typeof(CachedImage), false);
 
 		/// <summary>
-		/// If set to <c>true</c> DownsampleWidth and DownsampleHeight properties 
+		/// If set to <c>true</c> DownsampleWidth and DownsampleHeight properties
 		/// will use density independent pixels for downsampling
 		/// </summary>
 		/// <value><c>true</c> if downsample use dip units; otherwise, <c>false</c>.</value>
@@ -256,11 +256,11 @@ namespace FFImageLoading.Forms
 		{
 			get
 			{
-				return (bool)GetValue(DownsampleUseDipUnitsProperty); 
+				return (bool)GetValue(DownsampleUseDipUnitsProperty);
 			}
 			set
 			{
-				SetValue(DownsampleUseDipUnitsProperty, value); 
+				SetValue(DownsampleUseDipUnitsProperty, value);
 			}
 		}
 
@@ -276,11 +276,11 @@ namespace FFImageLoading.Forms
 		{
 			get
 			{
-				return (TimeSpan)GetValue(CacheDurationProperty); 
+				return (TimeSpan)GetValue(CacheDurationProperty);
 			}
 			set
 			{
-				SetValue(CacheDurationProperty, value); 
+				SetValue(CacheDurationProperty, value);
 			}
 		}
 
@@ -296,11 +296,11 @@ namespace FFImageLoading.Forms
         {
             get
             {
-                return (Work.LoadingPriority)GetValue(LoadingPriorityProperty); 
+                return (Work.LoadingPriority)GetValue(LoadingPriorityProperty);
             }
             set
             {
-                SetValue(LoadingPriorityProperty, value); 
+                SetValue(LoadingPriorityProperty, value);
             }
         }
 
@@ -318,11 +318,11 @@ namespace FFImageLoading.Forms
 		{
 			get
 			{
-				return (bool?)GetValue(TransparencyEnabledProperty); 
+				return (bool?)GetValue(TransparencyEnabledProperty);
 			}
 			set
 			{
-				SetValue(TransparencyEnabledProperty, value); 
+				SetValue(TransparencyEnabledProperty, value);
 			}
 		}
 
@@ -359,11 +359,11 @@ namespace FFImageLoading.Forms
 		{
 			get
 			{
-				return (bool?)GetValue(FadeAnimationEnabledProperty); 
+				return (bool?)GetValue(FadeAnimationEnabledProperty);
 			}
 			set
 			{
-				SetValue(FadeAnimationEnabledProperty, value); 
+				SetValue(FadeAnimationEnabledProperty, value);
 			}
 		}
 
@@ -380,11 +380,11 @@ namespace FFImageLoading.Forms
 		{
 			get
 			{
-				return (ImageSource)GetValue(LoadingPlaceholderProperty); 
+				return (ImageSource)GetValue(LoadingPlaceholderProperty);
 			}
 			set
 			{
-				SetValue(LoadingPlaceholderProperty, value); 
+				SetValue(LoadingPlaceholderProperty, value);
 			}
 		}
 
@@ -401,11 +401,11 @@ namespace FFImageLoading.Forms
 		{
 			get
 			{
-				return (ImageSource)GetValue(ErrorPlaceholderProperty); 
+				return (ImageSource)GetValue(ErrorPlaceholderProperty);
 			}
 			set
 			{
-				SetValue(ErrorPlaceholderProperty, value); 
+				SetValue(ErrorPlaceholderProperty, value);
 			}
 		}
 
@@ -443,11 +443,11 @@ namespace FFImageLoading.Forms
 		{
 			get
 			{
-                return (List<Work.ITransformation>)GetValue(TransformationsProperty); 
+                return (List<Work.ITransformation>)GetValue(TransformationsProperty);
 			}
 			set
 			{
-				SetValue(TransformationsProperty, value); 
+				SetValue(TransformationsProperty, value);
 			}
 		}
 
@@ -488,48 +488,88 @@ namespace FFImageLoading.Forms
 			base.OnBindingContextChanged();
 		}
 
-		protected override SizeRequest OnMeasure(double widthConstraint, double heightConstraint)
-		{
-			SizeRequest desiredSize = base.OnMeasure(double.PositiveInfinity, double.PositiveInfinity);
+              protected override SizeRequest OnMeasure(double widthConstraint, double heightConstraint)
+              {
+                     SizeRequest desiredSize = base.OnMeasure(double.PositiveInfinity, double.PositiveInfinity);
 
-			double desiredWidth = desiredSize.Request.Width;
-			double desiredHeight = desiredSize.Request.Height;
+                     double desiredWidth = desiredSize.Request.Width;
+                     double desiredHeight = desiredSize.Request.Height;
 
-			if (desiredWidth == 0 || desiredHeight == 0)
-				return new SizeRequest(new Size(0, 0));
+                     if (desiredWidth == 0 || desiredHeight == 0)
+                            return new SizeRequest(new Size(0, 0));
 
 
-			if (double.IsPositiveInfinity(widthConstraint) && double.IsPositiveInfinity(heightConstraint))
-			{
-				return new SizeRequest(new Size(desiredWidth, desiredHeight));
-			}
+                     if (double.IsPositiveInfinity(widthConstraint) && double.IsPositiveInfinity(heightConstraint))
+                     {
+                            return new SizeRequest(new Size(desiredWidth, desiredHeight));
+                     }
 
-			if (double.IsPositiveInfinity(widthConstraint))
-			{
-				double factor = heightConstraint / desiredHeight;
-				return new SizeRequest(new Size(desiredWidth * factor, desiredHeight * factor));
-			}
+                     if (double.IsPositiveInfinity(widthConstraint))
+                     {
+                            double factor = heightConstraint / desiredHeight;
+                            return new SizeRequest(new Size(desiredWidth * factor, desiredHeight * factor));
+                     }
 
-			if (double.IsPositiveInfinity(heightConstraint))
-			{
-				double factor = widthConstraint / desiredWidth;
-				return new SizeRequest(new Size(desiredWidth * factor, desiredHeight * factor));
-			}
+                     if (double.IsPositiveInfinity(heightConstraint))
+                     {
+                            double factor = widthConstraint / desiredWidth;
+                            return new SizeRequest(new Size(desiredWidth * factor, desiredHeight * factor));
+                     }
 
-			double fitsWidthRatio = widthConstraint / desiredWidth;
-			double fitsHeightRatio = heightConstraint / desiredHeight;
-			double ratioFactor = Math.Min(fitsWidthRatio, fitsHeightRatio);
+                     double desiredAspect = desiredSize.Request.Width / desiredSize.Request.Height;
+                     double constraintAspect = widthConstraint / heightConstraint;
+                     double width = desiredWidth;
+                     double height = desiredHeight;
 
-			return new SizeRequest(new Size(desiredWidth * ratioFactor, desiredHeight * ratioFactor));
-		}
+                     if (constraintAspect > desiredAspect)
+                     {
+                            // constraint area is proportionally wider than image
+                            switch (Aspect)
+                            {
+                                   case Aspect.AspectFit:
+                                   case Aspect.AspectFill:
+                                          height = Math.Min(desiredHeight, heightConstraint);
+                                          width = desiredWidth * (height / desiredHeight);
+                                          break;
+                                   case Aspect.Fill:
+                                          width = Math.Min(desiredWidth, widthConstraint);
+                                          height = desiredHeight * (width / desiredWidth);
+                                          break;
+                            }
+                     }
+                     else if (constraintAspect < desiredAspect)
+                     {
+                            // constraint area is proportionally taller than image
+                            switch (Aspect)
+                            {
+                                   case Aspect.AspectFit:
+                                   case Aspect.AspectFill:
+                                          width = Math.Min(desiredWidth, widthConstraint);
+                                          height = desiredHeight * (width / desiredWidth);
+                                          break;
+                                   case Aspect.Fill:
+                                          height = Math.Min(desiredHeight, heightConstraint);
+                                          width = desiredWidth * (height / desiredHeight);
+                                          break;
+                            }
+                     }
+                     else
+                     {
+                            // constraint area is same aspect as image
+                            width = Math.Min(desiredWidth, widthConstraint);
+                            height = desiredHeight * (width / desiredWidth);
+                     }
 
-		public void SetIsLoading(bool isLoading)
+                     return new SizeRequest(new Size(width, height));
+              }
+
+              public void SetIsLoading(bool isLoading)
 		{
 			SetValue(IsLoadingPropertyKey, isLoading);
 		}
 
 		internal Action InternalReloadImage;
-			
+
 		/// <summary>
 		/// Reloads the image.
 		/// </summary>
@@ -548,13 +588,13 @@ namespace FFImageLoading.Forms
         /// </summary>
 		public void Cancel()
 		{
-			if (InternalCancel != null) 
+			if (InternalCancel != null)
 			{
 				InternalCancel();
 			}
 		}
 
-		internal Func<GetImageAsJpgArgs, Task<byte[]>> InternalGetImageAsJPG; 
+		internal Func<GetImageAsJpgArgs, Task<byte[]>> InternalGetImageAsJPG;
 
 		/// <summary>
 		/// Gets the image as JPG.
@@ -594,7 +634,7 @@ namespace FFImageLoading.Forms
 		/// </summary>
 		/// <param name="source">Image source.</param>
 		/// <param name="cacheType">Cache type.</param>
-		/// <param name = "removeSimilar">If set to <c>true</c> removes all image cache variants 
+		/// <param name = "removeSimilar">If set to <c>true</c> removes all image cache variants
 		/// (downsampling and transformations variants)</param>
 		public static async Task InvalidateCache(ImageSource source, Cache.CacheType cacheType, bool removeSimilar = false)
 		{
@@ -608,7 +648,7 @@ namespace FFImageLoading.Forms
 
             var embResourceSource = source as EmbeddedResourceImageSource;
             if (embResourceSource != null)
-              await ImageService.Instance.InvalidateCacheEntryAsync(embResourceSource.Uri.OriginalString, cacheType, removeSimilar).ConfigureAwait(false);  
+              await ImageService.Instance.InvalidateCacheEntryAsync(embResourceSource.Uri.OriginalString, cacheType, removeSimilar).ConfigureAwait(false);
 		}
 
 		/// <summary>
@@ -616,7 +656,7 @@ namespace FFImageLoading.Forms
 		/// </summary>
 		/// <param name="source">Image key.</param>
 		/// <param name="cacheType">Cache type.</param>
-		/// <param name = "removeSimilar">If set to <c>true</c> removes all image cache variants 
+		/// <param name = "removeSimilar">If set to <c>true</c> removes all image cache variants
 		/// (downsampling and transformations variants)</param>
 		public static Task InvalidateCache(string key, Cache.CacheType cacheType, bool removeSimilar = false)
 		{
@@ -643,15 +683,15 @@ namespace FFImageLoading.Forms
 		{
 			get
 			{
-				return (ICommand)GetValue(SuccessCommandProperty); 
+				return (ICommand)GetValue(SuccessCommandProperty);
 			}
 			set
 			{
-				SetValue(SuccessCommandProperty, value); 
+				SetValue(SuccessCommandProperty, value);
 			}
 		}
 
-		internal void OnSuccess(CachedImageEvents.SuccessEventArgs e) 
+		internal void OnSuccess(CachedImageEvents.SuccessEventArgs e)
 		{
 			var handler = Success;
 			if (handler != null) handler(this, e);
@@ -681,15 +721,15 @@ namespace FFImageLoading.Forms
 		{
 			get
 			{
-				return (ICommand)GetValue(ErrorCommandProperty); 
+				return (ICommand)GetValue(ErrorCommandProperty);
 			}
 			set
 			{
-				SetValue(ErrorCommandProperty, value); 
+				SetValue(ErrorCommandProperty, value);
 			}
 		}
 
-		internal void OnError(CachedImageEvents.ErrorEventArgs e) 
+		internal void OnError(CachedImageEvents.ErrorEventArgs e)
 		{
 			var handler = Error;
 			if (handler != null) handler(this, e);
@@ -719,15 +759,15 @@ namespace FFImageLoading.Forms
 		{
 			get
 			{
-				return (ICommand)GetValue(FinishCommandProperty); 
+				return (ICommand)GetValue(FinishCommandProperty);
 			}
 			set
 			{
-				SetValue(FinishCommandProperty, value); 
+				SetValue(FinishCommandProperty, value);
 			}
 		}
 
-		internal void OnFinish(CachedImageEvents.FinishEventArgs e) 
+		internal void OnFinish(CachedImageEvents.FinishEventArgs e)
 		{
 			var handler = Finish;
 			if (handler != null) handler(this, e);
@@ -870,7 +910,7 @@ namespace FFImageLoading.Forms
         }
 
         /// <summary>
-        /// Setups the on before image loading. 
+        /// Setups the on before image loading.
         /// You can add additional logic here to configure image loader settings before loading
         /// </summary>
         /// <param name="imageLoader">Image loader.</param>

